pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,474 @@
+	// was injected for too long, the cockroach process will detect a disk stall
+	// and exit. This deletes the cockroach service and there is no need to
+	// unlimit anything. Instead, restart the node if RestartNodes is true.
+	err = s.setThroughput(ctx, l, stallTypes, throughput{limited: false}, nodes, cockroachIOController)
+	return forEachNode(diskStallArgs.Nodes, func(n install.Nodes) error {
+		err = s.Run(ctx, l, n, "cat", cockroachIOController)
+		if err != nil && diskStallArgs.RestartNodes {
+			l.Printf("failed to log cgroup bandwidth limits, assuming n%d exited and restarting: %v", n, err)
+			return s.StartNodes(ctx, l, n)
+		}
+		return nil
+	})
+}
+
+func (s *CGroupDiskStaller) WaitForFailureToPropagate(
+	ctx context.Context, l *logger.Logger, args FailureArgs,
+) error {
+	diskStallArgs := args.(DiskStallArgs)
+	if diskStallArgs.StallWrites {
+		// If writes are stalled, we expect the disk stall detection to kick in
+		// and kill the node.
+		return forEachNode(diskStallArgs.Nodes, func(n install.Nodes) error {
+			return s.WaitForSQLNodeDeath(ctx, l, n, time.Minute)
+		})
+	}
+	return nil
+}
+
+func (s *CGroupDiskStaller) WaitForFailureToRecover(
+	ctx context.Context, l *logger.Logger, args FailureArgs,
+) error {
+	nodes := args.(DiskStallArgs).Nodes
+	return forEachNode(nodes, func(n install.Nodes) error {
+		return s.WaitForSQLReady(ctx, l, n, time.Minute)
+	})
+}
+
+type throughput struct {
+	limited        bool
+	bytesPerSecond string
+}
+
+type bandwidthType int8
+
+const (
+	readBandwidth bandwidthType = iota
+	writeBandwidth
+)
+
+func (rw bandwidthType) cgroupV2BandwidthProp() string {
+	switch rw {
+	case readBandwidth:
+		return "rbps"
+	case writeBandwidth:
+		return "wbps"
+	default:
+		panic("unreachable")
+	}
+}
+
+func (s *CGroupDiskStaller) setThroughput(
+	ctx context.Context,
+	l *logger.Logger,
+	readOrWrite []bandwidthType,
+	bw throughput,
+	nodes install.Nodes,
+	cockroachIOController string,
+) error {
+	maj, min, err := s.DiskDeviceMajorMinor(ctx, l)
+	if err != nil {
+		return err
+	}
+
+	var limits []string
+	for _, rw := range readOrWrite {
+		bytesPerSecondStr := "max"
+		if bw.limited {
+			bytesPerSecondStr = bw.bytesPerSecond
+		}
+		limits = append(limits, fmt.Sprintf("%s=%s", rw.cgroupV2BandwidthProp(), bytesPerSecondStr))
+	}
+	l.Printf("setting cgroup bandwith limits:\n%v", limits)
+
+	return s.Run(ctx, l, nodes, "sudo", "/bin/bash", "-c", fmt.Sprintf(
+		`'echo %d:%d %s > %s'`,
+		maj,
+		min,
+		strings.Join(limits, " "),
+		cockroachIOController,
+	))
+}
+
+func (s *CGroupDiskStaller) GetReadWriteBytes(
+	ctx context.Context, l *logger.Logger, node install.Nodes,
+) (int, int, error) {
+	maj, min, err := s.DiskDeviceMajorMinor(ctx, l)
+	if err != nil {
+		return 0, 0, err
+	}
+	// Check the number of bytes read and written to disk.
+	res, err := s.RunWithDetails(
+		ctx, l, node,
+		fmt.Sprintf(`grep -E '%d:%d' /sys/fs/cgroup/system.slice/io.stat |`, maj, min),
```

Comment:
It seems there are some lightweight metrics collectors, e.g., [1], which we could easily add as a scrape target. This could make it easier to debug; rather than sampling over discrete intervals ourselves, we'd have a more or less continuous view in Grafana.

[1] https://github.com/arianvp/cgroup-exporter/tree/main/collector